File: backend/src/modules/security/security.service.js
// ============================================
// Security Service
// Dashboard, issues, login history, alerts, 2FA toggle
// ============================================
const prisma = require('../../config/database');
const { isStrongPassword } = require('../../utils/helpers');

/**
 * Lấy dữ liệu Security Dashboard
 * Trả về overview bảo mật cho user
 */
async function getDashboard(userId) {
  const user = await prisma.user.findUnique({
    where: { id: userId },
    select: {
      id: true,
      name: true,
      email: true,
      phoneNumber: true,
      emailVerified: true,
      phoneVerified: true,
      twoFactorEnabled: true,
      role: true,
      failedLoginAttempts: true,
      lockedUntil: true,
      lastLoginAt: true,
      createdAt: true,
    },
  });

  if (!user) {
    throw Object.assign(new Error('User not found'), { statusCode: 404 });
  }

  // Đếm unread alerts
  const unreadAlerts = await prisma.securityAlert.count({
    where: { userId, read: false },
  });

  // Đếm login suspicious
  const suspiciousLogins = await prisma.loginHistory.count({
    where: { userId, suspicious: true },
  });

  // Lấy 5 login gần nhất
  const recentLogins = await prisma.loginHistory.findMany({
    where: { userId },
    orderBy: { createdAt: 'desc' },
    take: 5,
    select: {
      id: true,
      email: true,
      ipAddress: true,
      deviceName: true,
      success: true,
      suspicious: true,
      createdAt: true,
    },
  });

  // Tính Security Score
  const { score, issues } = calculateSecurityScore(user, suspiciousLogins);

  const recentAlerts = await prisma.securityAlert.findMany({
    where: { userId },
    orderBy: { createdAt: 'desc' },
    take: 5,
    select: {
      id: true,
      type: true,
      message: true,
      severity: true,
      read: true,
      createdAt: true,
    },
  });

  return {
    hasEmail: !!user.email,
    hasPhoneNumber: !!user.phoneNumber,
    accountVerified: !!user.emailVerified || !!user.phoneVerified,
    emailVerified: user.emailVerified,
    phoneVerified: user.phoneVerified,
    twoFactorEnabled: user.twoFactorEnabled,
    lastLoginAt: user.lastLoginAt,
    failedLoginAttempts: user.failedLoginAttempts,
    accountLocked: user.lockedUntil ? user.lockedUntil > new Date() : false,
    securityScore: score,
    recentAlerts,
    user: {
      id: user.id,
      name: user.name,
      email: user.email,
      phoneNumber: user.phoneNumber,
      role: user.role,
      createdAt: user.createdAt,
    },
    security: {
      emailVerified: user.emailVerified,
      phoneVerified: user.phoneVerified,
      hasEmail: !!user.email,
      hasPhoneNumber: !!user.phoneNumber,
      twoFactorEnabled: user.twoFactorEnabled,
      lastLogin: user.lastLoginAt,
      failedLoginAttempts: user.failedLoginAttempts,
      accountLocked: user.lockedUntil ? user.lockedUntil > new Date() : false,
      lockedUntil: user.lockedUntil,
      securityScore: score,
      unreadAlerts,
      suspiciousLogins,
    },
    securityIssues: issues,
    recentLogins,
  };
}

/**
 * Tính Security Score (0-100)
 * Bắt đầu từ 100, trừ điểm theo từng vấn đề
 */
function calculateSecurityScore(user, suspiciousLogins) {
  let score = 100;
  const issues = [];

  // Email chưa xác minh: -30 điểm (nếu có email)
  if (user.email && !user.emailVerified) {
    score -= 30;
    issues.push({
      type: 'EMAIL_NOT_VERIFIED',
      title: 'Email not verified',
      description: 'Verify your email to secure your account',
      severity: 'HIGH',
      impact: -30,
    });
  }

  // Số điện thoại chưa xác minh: -25 điểm (nếu có phone)
  if (user.phoneNumber && !user.phoneVerified) {
    score -= 25;
    issues.push({
      type: 'PHONE_NOT_VERIFIED',
      title: 'Phone number not verified',
      description: 'Verify your phone number to secure your account',
      severity: 'HIGH',
      impact: -25,
    });
  }

  // Chưa thêm số điện thoại: -15 điểm
  if (!user.phoneNumber) {
    score -= 15;
    issues.push({
      type: 'NO_PHONE_NUMBER',
      title: 'Phone number missing',
      description: 'Add a phone number for fallback security option',
      severity: 'MEDIUM',
      impact: -15,
    });
  }

  // Chưa bật 2FA: -20 điểm
  if (!user.twoFactorEnabled) {
    score -= 20;
    issues.push({
      type: 'TWO_FACTOR_DISABLED',
      title: 'Two-factor authentication disabled',
      description: 'Enable 2FA for an extra layer of security',
      severity: 'MEDIUM',
      impact: -20,
    });
  }

  // Nhiều lần đăng nhập thất bại (>3): -15 điểm
  if (user.failedLoginAttempts > 3) {
    score -= 15;
    issues.push({
      type: 'MULTIPLE_FAILED_LOGINS',
      title: 'Multiple failed login attempts',
      description: `${user.failedLoginAttempts} failed login attempts detected`,
      severity: 'HIGH',
      impact: -15,
    });
  }

  // Có đăng nhập từ thiết bị lạ: -15 điểm
  if (suspiciousLogins > 0) {
    score -= 15;
    issues.push({
      type: 'SUSPICIOUS_LOGINS',
      title: 'Suspicious login detected',
      description: `${suspiciousLogins} login(s) from unrecognized devices`,
      severity: 'MEDIUM',
      impact: -15,
    });
  }

  // Tài khoản đang bị khóa: -20 điểm
  if (user.lockedUntil && user.lockedUntil > new Date()) {
    score -= 20;
    issues.push({
      type: 'ACCOUNT_LOCKED',
      title: 'Account is currently locked',
      description: 'Your account has been temporarily locked due to security concerns',
      severity: 'CRITICAL',
      impact: -20,
    });
  }

  // Đảm bảo score không âm
  score = Math.max(0, score);

  return { score, issues };
}

/**
 * Lấy danh sách security issues
 */
async function getSecurityIssues(userId) {
  const user = await prisma.user.findUnique({
    where: { id: userId },
  });

  if (!user) {
    throw Object.assign(new Error('User not found'), { statusCode: 404 });
  }

  const suspiciousLogins = await prisma.loginHistory.count({
    where: { userId, suspicious: true },
  });

  const { issues } = calculateSecurityScore(user, suspiciousLogins);

  return issues;
}

/**
 * Lấy lịch sử đăng nhập
 */
async function getLoginHistory(userId, page = 1, limit = 20) {
  const skip = (page - 1) * limit;

  const [history, total] = await Promise.all([
    prisma.loginHistory.findMany({
      where: { userId },
      orderBy: { createdAt: 'desc' },
      skip,
      take: limit,
      select: {
        id: true,
        email: true,
        ipAddress: true,
        userAgent: true,
        deviceName: true,
        success: true,
        reason: true,
        suspicious: true,
        createdAt: true,
      },
    }),
    prisma.loginHistory.count({ where: { userId } }),
  ]);

  return {
    history,
    pagination: {
      page,
      limit,
      total,
      totalPages: Math.ceil(total / limit),
    },
  };
}

/**
 * Lấy security alerts
 */
async function getSecurityAlerts(userId, page = 1, limit = 20) {
  const skip = (page - 1) * limit;

  const [alerts, total] = await Promise.all([
    prisma.securityAlert.findMany({
      where: { userId },
      orderBy: { createdAt: 'desc' },
      skip,
      take: limit,
      select: {
        id: true,
        type: true,
        message: true,
        severity: true,
        read: true,
        createdAt: true,
      },
    }),
    prisma.securityAlert.count({ where: { userId } }),
  ]);

  return {
    alerts,
    pagination: {
      page,
      limit,
      total,
      totalPages: Math.ceil(total / limit),
    },
  };
}

/**
 * Bật/tắt 2FA
 */
async function toggle2FA(userId, enabled) {
  const user = await prisma.user.findUnique({
    where: { id: userId },
  });

  if (!user) {
    throw Object.assign(new Error('User not found'), { statusCode: 404 });
  }

  // Yêu cầu email hoặc số điện thoại phải được verify trước khi bật 2FA
  if (!user.emailVerified && !user.phoneVerified && !user.twoFactorEnabled) {
    throw Object.assign(
      new Error('Please verify your email or phone number before enabling 2FA'),
      { statusCode: 400 }
    );
  }

  const newStatus = typeof enabled === 'boolean' ? enabled : !user.twoFactorEnabled;

  await prisma.user.update({
    where: { id: userId },
    data: { twoFactorEnabled: newStatus },
  });

  // Tạo security alert
  await prisma.securityAlert.create({
    data: {
      userId,
      type: newStatus ? 'TWO_FACTOR_ENABLED' : 'TWO_FACTOR_DISABLED',
      message: newStatus
        ? 'Two-factor authentication has been enabled'
        : 'Two-factor authentication has been disabled',
      severity: newStatus ? 'LOW' : 'MEDIUM',
    },
  });

  return {
    twoFactorEnabled: newStatus,
    message: newStatus
      ? '2FA has been enabled. You will need to enter OTP when logging in.'
      : '2FA has been disabled.',
  };
}

/**
 * Đánh dấu alert đã đọc
 */
async function markAlertAsRead(userId, alertId) {
  await prisma.securityAlert.updateMany({
    where: {
      id: alertId,
      userId,
    },
    data: { read: true },
  });

  return { message: 'Alert marked as read' };
}

module.exports = {
  getDashboard,
  getSecurityIssues,
  getLoginHistory,
  getSecurityAlerts,
  toggle2FA,
  markAlertAsRead,
};

File: flutter_app/lib/services/security_service.dart
import 'dart:math';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';

class SecurityService {
  final _storage = const FlutterSecureStorage();
  final FirebaseAuth _auth = FirebaseAuth.instance;

  // Cache in-memory
  bool _twoFactorEnabled = false;
  List<Map<String, dynamic>> _loginHistory = [];
  List<Map<String, dynamic>> _alerts = [];
  bool _initialized = false;

  Future<void> _initIfNeeded() async {
    if (_initialized) return;
    
    // Đọc trạng thái 2FA từ secure storage
    final tfaVal = await _storage.read(key: 'two_factor_enabled');
    _twoFactorEnabled = tfaVal == 'true';

    // Tạo sẵn một số dữ liệu lịch sử đăng nhập demo cho đẹp mắt
    _loginHistory = [
      {
        'id': 1,
        'email': _auth.currentUser?.email ?? 'user@example.com',
        'ipAddress': '192.168.1.100',
        'deviceName': 'Android Emulator',
        'success': true,
        'suspicious': false,
        'createdAt': DateTime.now().subtract(const Duration(hours: 2)).toIso8601String(),
      },
      {
        'id': 2,
        'email': _auth.currentUser?.email ?? 'user@example.com',
        'ipAddress': '14.232.84.102',
        'deviceName': 'Chrome Browser - Windows',
        'success': true,
        'suspicious': true,
        'reason': null,
        'createdAt': DateTime.now().subtract(const Duration(days: 1)).toIso8601String(),
      },
      {
        'id': 3,
        'email': _auth.currentUser?.email ?? 'user@example.com',
        'ipAddress': '192.168.1.15',
        'deviceName': 'iPhone 15 Pro',
        'success': false,
        'reason': 'Sai mật khẩu',
        'suspicious': false,
        'createdAt': DateTime.now().subtract(const Duration(days: 2)).toIso8601String(),
      }
    ];

    // Tạo sẵn cảnh báo bảo mật demo
    _alerts = [
      if (_twoFactorEnabled == false)
        {
          'id': 101,
          'type': 'TWO_FACTOR_DISABLED',
          'message': 'Xác thực 2 bước (2FA) đang bị tắt. Hãy bật để tăng bảo mật.',
          'severity': 'MEDIUM',
          'read': false,
          'createdAt': DateTime.now().subtract(const Duration(days: 1)).toIso8601String(),
        },
      {
        'id': 102,
        'type': 'SUSPICIOUS_LOGIN',
        'message': 'Phát hiện đăng nhập lạ từ địa chỉ IP 14.232.84.102 (Windows Device)',
        'severity': 'MEDIUM',
        'read': true,
        'createdAt': DateTime.now().subtract(const Duration(days: 1)).toIso8601String(),
      }
    ];

    _initialized = true;
  }

  /// Ghi nhận log đăng nhập khi user login thành công
  Future<void> recordLoginLog(String deviceName) async {
    await _initIfNeeded();
    final user = _auth.currentUser;
    final email = user?.email ?? 'unknown@example.com';
    final timestamp = DateTime.now().toIso8601String();

    // Check xem thiết bị này đã từng đăng nhập thành công chưa
    final bool isKnown = _loginHistory.any((log) => log['deviceName'] == deviceName && log['success'] == true);
    final bool isSuspicious = _loginHistory.isNotEmpty && !isKnown;

    final newLog = {
      'id': _loginHistory.length + 1,
      'email': email,
      'ipAddress': '192.168.1.50',
      'deviceName': deviceName,
      'success': true,
      'suspicious': isSuspicious,
      'createdAt': timestamp,
    };

    _loginHistory.insert(0, newLog);

    if (isSuspicious) {
      final newAlert = {
        'id': _alerts.length + 101,
        'type': 'SUSPICIOUS_LOGIN',
        'message': 'Đăng nhập lạ từ thiết bị mới phát hiện: $deviceName',
        'severity': 'HIGH',
        'read': false,
        'createdAt': timestamp,
      };
      _alerts.insert(0, newAlert);
    }
  }

  /// Lấy dữ liệu Security Dashboard
  Future<Map<String, dynamic>> getDashboard() async {
    await _initIfNeeded();
    final user = _auth.currentUser;

    final bool emailVerified = user?.emailVerified ?? false;
    
    // Tính toán điểm bảo mật (tổng 100)
    int score = 100;
    final List<Map<String, dynamic>> issues = [];

    if (!emailVerified) {
      score -= 30;
      issues.add({
        'type': 'EMAIL_NOT_VERIFIED',
        'title': 'Email chưa được xác minh',
        'description': 'Xác minh email để nâng cao tính bảo mật cho tài khoản của bạn.',
        'severity': 'HIGH',
        'impact': -30,
      });
    }

    if (!_twoFactorEnabled) {
      score -= 20;
      issues.add({
        'type': 'TWO_FACTOR_DISABLED',
        'title': 'Chưa bật xác thực 2 bước (2FA)',
        'description': 'Kích hoạt xác thực 2 lớp để yêu cầu kiểm tra liên kết qua email mỗi khi đăng nhập.',
        'severity': 'MEDIUM',
        'impact': -20,
      });
    }

    final suspiciousCount = _loginHistory.where((log) => log['suspicious'] == true).length;
    if (suspiciousCount > 0) {
      score -= 15;
      issues.add({
        'type': 'SUSPICIOUS_LOGINS',
        'title': 'Phát hiện đăng nhập nghi vấn',
        'description': 'Có $suspiciousCount phiên đăng nhập từ các thiết bị chưa nhận dạng.',
        'severity': 'MEDIUM',
        'impact': -15,
      });
    }

    score = score.clamp(0, 100);

    return {
      'hasEmail': user?.email != null,
      'hasPhoneNumber': user?.phoneNumber != null,
      'accountVerified': emailVerified,
      'emailVerified': emailVerified,
      'phoneVerified': user?.phoneNumber != null,
      'twoFactorEnabled': _twoFactorEnabled,
      'lastLoginAt': _loginHistory.isNotEmpty ? _loginHistory.first['createdAt'] : null,
      'failedLoginAttempts': 0,
      'accountLocked': false,
      'securityScore': score,
      'recentAlerts': _alerts.take(5).toList(),
      'user': {
        'id': user?.uid.hashCode ?? 0,
        'name': user?.displayName ?? 'Firebase User',
        'email': user?.email ?? '',
        'role': 'USER',
        'createdAt': user?.metadata.creationTime?.toIso8601String() ?? '',
      },
      'security': {
        'emailVerified': emailVerified,
        'phoneVerified': user?.phoneNumber != null,
        'hasEmail': user?.email != null,
        'hasPhoneNumber': user?.phoneNumber != null,
        'twoFactorEnabled': _twoFactorEnabled,
        'lastLogin': _loginHistory.isNotEmpty ? _loginHistory.first['createdAt'] : null,
        'failedLoginAttempts': 0,
        'accountLocked': false,
        'securityScore': score,
        'unreadAlerts': _alerts.where((a) => a['read'] == false).length,
        'suspiciousLogins': suspiciousCount,
      },
      'securityIssues': issues,
      'recentLogins': _loginHistory.take(5).toList(),
    };
  }

  /// Lấy danh sách các sự cố bảo mật hiện tại
  Future<List<dynamic>> getIssues() async {
    final dashboard = await getDashboard();
    return dashboard['securityIssues'] ?? [];
  }

  /// Lấy lịch sử đăng nhập phân trang
  Future<Map<String, dynamic>> getLoginHistory({
    int page = 1,
    int limit = 10,
  }) async {
    await _initIfNeeded();
    final total = _loginHistory.length;
    final startIndex = (page - 1) * limit;
    final endIndex = (startIndex + limit).clamp(0, total);
    
    final pagedItems = _loginHistory.sublist(startIndex, endIndex);

    return {
      'history': pagedItems,
      'pagination': {
        'page': page,
        'limit': limit,
        'total': total,
        'totalPages': (total / limit).ceil(),
      }
    };
  }

  /// Lấy danh sách cảnh báo bảo mật
  Future<List<dynamic>> getAlerts() async {
    await _initIfNeeded();
    return _alerts;
  }

  /// Bật/Tắt xác thực 2 bước (2FA)
  Future<Map<String, dynamic>> toggle2FA(bool enabled) async {
    await _initIfNeeded();
    _twoFactorEnabled = enabled;
    await _storage.write(key: 'two_factor_enabled', value: enabled.toString());

    // Thêm cảnh báo tương ứng
    final timestamp = DateTime.now().toIso8601String();
    final newAlert = {
      'id': _alerts.length + 101,
      'type': enabled ? 'TWO_FACTOR_ENABLED' : 'TWO_FACTOR_DISABLED',
      'message': enabled 
          ? 'Xác thực 2 bước (2FA) đã được bật thành công!' 
          : 'Cảnh báo: Xác thực 2 bước (2FA) đã bị tắt.',
      'severity': enabled ? 'LOW' : 'MEDIUM',
      'read': false,
      'createdAt': timestamp,
    };
    _alerts.insert(0, newAlert);

    return {
      'twoFactorEnabled': _twoFactorEnabled,
      'message': enabled
          ? 'Đã bật 2FA thành công. Bạn sẽ cần xác thực email mỗi khi đăng nhập.'
          : 'Đã tắt xác thực 2 bước.',
    };
  }

  /// Kiểm tra xem 2FA đã được xác thực cho phiên hiện tại chưa
  Future<bool> isTwoFactorVerifiedForSession() async {
    final verified = await _storage.read(key: 'two_factor_session_verified');
    return verified == 'true';
  }

  /// Cập nhật trạng thái xác thực 2FA cho phiên hiện tại
  Future<void> setTwoFactorVerified(bool verified) async {
    await _storage.write(key: 'two_factor_session_verified', value: verified.toString());
  }

  String? _currentOTP;
  DateTime? _otpExpireTime;

  /// Sinh mã OTP 2FA ngẫu nhiên 6 chữ số
  String generate2FAOTP() {
    final rand = Random();
    final otp = (100000 + rand.nextInt(900000)).toString();
    _currentOTP = otp;
    _otpExpireTime = DateTime.now().add(const Duration(minutes: 5));
    return otp;
  }

  /// Xác thực mã OTP 2FA
  bool verifyOTP(String code) {
    if (_currentOTP == null || _otpExpireTime == null) return false;
    if (DateTime.now().isAfter(_otpExpireTime!)) {
      _currentOTP = null;
      _otpExpireTime = null;
      return false;
    }
    final isValid = _currentOTP == code.trim();
    if (isValid) {
      _currentOTP = null; // Xoá mã sau khi xác thực thành công
      _otpExpireTime = null;
    }
    return isValid;
  }

  /// Xóa trạng thái xác thực 2FA của phiên (khi logout)
  Future<void> clearTwoFactorSession() async {
    await _storage.delete(key: 'two_factor_session_verified');
    _currentOTP = null;
    _otpExpireTime = null;
  }
}

final securityService = SecurityService();
